File: src/main/scala/scalatest/Kafka_Load.py
from confluent_kafka import Producer, Consumer
import json
import time
import random
import happybase

# Kafka configurations
kafka_config = {
    'bootstrap.servers': '<kafka_broker>',
    'group.id': 'my_group',
}

# HBase configurations
hbase_host = '<hbase_host>'
hbase_table_name = 'my_table'

# Function to produce data to Kafka topic
def produce_data():
    producer = Producer(kafka_config)
    while True:
        records = [{'value': {'id': i, 'data': random.randint(1, 100)}} for i in range(1000)]
        for record in records:
            producer.produce('<kafka_topic>', json.dumps(record))
        producer.flush()
        time.sleep(5)

# Function to consume data from Kafka topic and store in HBase
def consume_and_store_data():
    connection = happybase.Connection(hbase_host)
    table = connection.table(hbase_table_name)
    consumer = Consumer(kafka_config)
    consumer.subscribe(['<kafka_topic>'])
    while True:
        msg = consumer.poll(timeout=10)
        if msg is None:
            continue
        if msg.error():
            print("Consumer error: {}".format(msg.error()))
            continue
        try:
            data = json.loads(msg.value().decode('utf-8'))
            row_key = str(data['id'])
            table.put(row_key, {'cf:data': str(data['data'])})
            print("Data stored in HBase:", data)
        except Exception as e:
            print("Error processing message:", e)

# Run producer and consumer concurrently
if __name__ == "__main__":
    import threading
    producer_thread = threading.Thread(target=produce_data)
    consumer_thread = threading.Thread(target=consume_and_store_data)
    producer_thread.start()
    consumer_thread.start()

File: src/main/scala/scalatest/Kafka_Load.scala
import scala.io.Source
import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions._
import org.apache.hadoop.hbase.util.Bytes
import org.apache.hadoop.hbase.client.{Connection, ConnectionFactory, Put, Admin}
import org.apache.hadoop.hbase.{HBaseConfiguration, TableName, HColumnDescriptor, HTableDescriptor}
import java.util.concurrent.TimeUnit
import scala.util.Random

object KafkaHBaseProcessor extends App {

  // Function to fetch data from API
  def fetchDataFromAPI(apiUrl: String): String = {
    Source.fromURL(apiUrl).mkString
  }

  // Function to publish data to Kafka
  def publishToKafka(data: String, topic: String, kafkaServers: String, spark: SparkSession): Unit = {
    val jsonData = spark.read.json(Seq(data).toDS())
    jsonData.selectExpr("to_json(struct(*)) AS value").write.format("kafka")
      .option("kafka.bootstrap.servers", kafkaServers)
      .option("topic", topic)
      .save()
  }

  // Function to consume data from Kafka and load it into HBase
  def consumeAndLoadToHBase(topic: String, kafkaServers: String, spark: SparkSession): Unit = {
    val df = spark.read.format("kafka")
      .option("kafka.bootstrap.servers", kafkaServers)
      .option("subscribe", topic)
      .option("startingOffsets", "earliest")
      .load()

    val messages = df.selectExpr("CAST(value AS STRING) as message")

    // Configure HBase connection
    val hbaseConf = HBaseConfiguration.create()
    hbaseConf.set("hbase.zookeeper.quorum", "zookeeper-server1,zookeeper-server2,zookeeper-server3")
    hbaseConf.set("hbase.zookeeper.property.clientPort", "2181")
    hbaseConf.set("zookeeper.znode.parent", "/hbase")

    // Establish HBase connection
    val connection = ConnectionFactory.createConnection(hbaseConf)

    // Prepare HBase table for storing the messages
    val tableName = TableName.valueOf(topic)
    val columnFamilyName = "cf"
    val admin = connection.getAdmin

    // Check if table exists, if not, create it
    if (!admin.tableExists(tableName)) {
      val tableDescriptor = new HTableDescriptor(tableName)
      val columnDescriptor = new HColumnDescriptor(columnFamilyName)
      tableDescriptor.addFamily(columnDescriptor)
      admin.createTable(tableDescriptor)
      println(s"Table $tableName created.")
    }

    // Get the HBase table
    val table = connection.getTable(tableName)

    // Function to generate unique row keys for messages
    def generateUniqueRowKey(): String = {
      val currentTime = System.currentTimeMillis()
      val random = new Random()
      s"$currentTime-${random.nextInt(10000)}"
    }

    // Insert each Kafka message into HBase
    messages.collect().take(1000).foreach { message =>
      val rowKey = generateUniqueRowKey()
      val put = new Put(Bytes.toBytes(rowKey))
      put.addColumn(Bytes.toBytes(columnFamilyName), Bytes.toBytes("data"), Bytes.toBytes(message.getString(0)))
      table.put(put)
    }

    // Print summary of operations
    println("Finished loading data to HBase.")

    // Close HBase table and connection to clean up resources
    table.close()
    connection.close()
  }

  // Kafka servers configuration
  val kafkaServers = "kafka-server1:9092,kafka-server2:9092,kafka-server3:9092"
  val topic = "insurance_claims_topic"

  // Initialize Spark session
  val spark = SparkSession.builder()
    .appName("KafkaHBaseProcessor")
    .getOrCreate()

  // Produce data every 5 seconds
  while (true) {
    val data = fetchDataFromAPI("http://18.133.73.36:5001/insurance_claims1")
    publishToKafka(data, topic, kafkaServers, spark)
    println(s"Published data to Kafka at ${java.time.LocalDateTime.now()}")
    TimeUnit.SECONDS.sleep(5)
  }

  // Consume data and load into HBase every 10 seconds
  while (true) {
    consumeAndLoadToHBase(topic, kafkaServers, spark)
    println(s"Loaded data from Kafka to HBase at ${java.time.LocalDateTime.now()}")
    TimeUnit.SECONDS.sleep(10)
  }

  // Stop Spark session
  spark.stop()
}
